Autograder

Sjtim opened the pull request:
npm install

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

Summary:

Autograder Feature for the Evaluationprocess. A teacher can configure a GradeDefinition which will be applied on their respective EvaluationStep. Each Step has its own GradeDefinition to calculate Points dependent on a points-cap (max-points) and type of mistake (minor, major, critical).
All reached points of an Evaluation will be put together to calculate a Grade for Student Submissions.
Under any circumstances a teacher can adjust given points.
If a students submission fails in any evaluationstep the teacher has to take action at triggers a calculation if he adjusts points manually. This also occures if he adjusts points of an evaluation which already has a grade applied.

- [x] I have performed a self-review of my own code
- [x] I have run the auto code formatting scripts <!-- npm run fix, gradle spotlessApply -->
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I updated the `CHANGELOG.md`
- [ ] I have added tests that prove my fix is effective or that my feature works

Sjtim committed:
4 more students require an AuthorizationTest.kt

Sjtim committed:
Added autograder schema to existing database structure.
yaml files generated with inbuild script.

Sjtim committed:
Init GradeDefinition Entity which holds a foreign key of the already existing EvaluationStepDefinition
The GradeDefinition defines multiply PointsOfEvaluationStep(s) how to calculate their Points Fields.

Sjtim committed:
An Alias Name Generator Class plus a Json file which contains animal names.

It has a public function called getRandomAnimal(user:User) which adds a random Animal name to the provided user

Has an init method apply one of those names with RNG on all existing users

Sjtim committed:
UserAlias Entity + Spring CRUD Interface which connects to the existing User Entity.
Adds the option for an UserAlias name.

required for feature of anonymous scoreboard.

Sjtim committed:
Added the UserAlias Entity.

Sjtim committed:
GradeDefinition CRUD interface with some generic functions for DB access

Sjtim committed:
Grade Entity which is connection to an Evaluation Entity.
It is also part of the Answer Entity and holds n PointsOfEvaluationStep Entities.

Provides a Grade for each evaluation and as a Set in the Answer Entity it help to pick the best of the set.

Sjtim committed:
GradeRepository interface has CRUD and generic functions to access the DB for the Grade Entity

Sjtim committed:
PointsOfEvaluationStep Entity + PointsOfEvaluationStepRepository provides Entity and interface for db access

A PointsOfEvaluationStep has a OneToOne relation to an EvaluationStep. This will provide points depending on the EvaluationStep Result and its Set of Feedbacks for a possible Grade.

Can later be applied to an Grade if all requirements are met.

Sjtim committed:
The GradeService handles all access from Grades to the Database.

It has functions to calculate / recalculate a grade by different sources.

Sjtim committed:
The GradeDefinitionServices handles all access to the Database related to a GradeDefinition.

Sjtim committed:
PointsOfEvaluationStepService controls all acces to the database of the PointsOfEvaluationStep entity.
Futhermore it has functions to perform calculation operations for points reached in an EvaluationStep.

Sjtim committed:
Has been modified in afterStep function.
After the evaluationStepStatus has been updated we can start the calculation process for possible points if autograding is enabled.

Sjtim committed:
Added functions to add a GradeDefinition and PointsOfEvaluationStep to EvaluationStep entities.

if successful and requirements are met a calculation will perform.

Sjtim committed:
Added a Set of Grades to the Answer Entity. Required to find the best grade of an answer.
It will also be possible to find the best over the connected EvaluationSteps. Cascading might ramp up the load and this will be more handy.

Sjtim committed:
Added a GradeDefinition to the EvaluationStepDefinition. Required to define one for grades to every possible Evaluation.

Sjtim committed:
Added PointsOfEvaluation and GradeDefinition to the EvaluationStep Entity. Required to get the connection between a definition defined in a task and possible archived points in this EvaluationStep.

Sjtim committed:
Every Evaluation might have a Grade.

Sjtim committed:
Added function to find All Answers by a Submission Id

Sjtim committed:
Added function to find an Evaluation by one of its EvaluationSteps.

Sjtim committed:
enhanced the Body and added some generic functions

Sjtim committed:
added some function to access two cases which are required to get the flawed and errored feedbacks

Sjtim committed:
Added a couple of students for evaluation

Sjtim committed:
Added UserAliasService to access the UserAlias Entity. Contains base crud functions

Sjtim committed:
addBuiltInEvaluationSteps function modified to add a GradeDefinition to all evaluationStepDefinition.

Sjtim committed:
Outcommented the Commandlinerunner for inbuild Database. Switched to Postgresql

Sjtim committed:
Added the findAllBySubmissionId function to access its respective function in answerRepository.

Sjtim committed:
Added the function findAssignmentById to access the recently generic generated function in assignmentRepository.

Sjtim committed:
Added a GradeDefinition to the existing EvaluationStepDefinitionDto to gain access on Frontend of its values.
Added a PointsOfEvaluationStep to the existing EvaluationStepDto to gain access on Frontend of its values.
Modified the createEvaluationStepDefinition Mutation and added a gradeDefinition if its missing.

Sjtim committed:
AutograderApi to query/mutate all cases
Provides Dtos for Query, input Dtos for mutations and their respective functions for frontend access.

Sjtim committed:
Config files. Autogenerated and edited some values in tsconfig.json to develop

Sjtim committed:
enhanced trees

Sjtim committed:
Added Autograder Querys

Sjtim committed:
Added Autograder Mutations

Sjtim committed:
changed process order for calculation

Sjtim committed:
fixed couple of bugs with grade calculation.

Sjtim committed:
findById added

Sjtim committed:
Added API to turn on/off a scoreboard.

Sjtim committed:
Split the gradeDefinition Input. Now to turn it on/off and change their values are two different Querys / Mutations

Sjtim committed:
Added Nullcheck to PointsOfEvaluationStep. Might not exists.

Sjtim committed:
Added Boolean Value to turn/off the Scoreboard

Sjtim committed:
Added Component to mutate the nickname

Sjtim committed:
Added a ScoreboardTable + Page.

Sjtim committed:
UseGetGrade Hook added.

Sjtim committed:
usePointsOFEvaluationStep Hook added.

Sjtim committed:
gradeDefinitionMax modified

Sjtim committed:
GradeDefinitionMax modified and added scoreboard boolean to assignment

Sjtim committed:
Added mutation for scoreboard boolean to turn on/off the scoreboard

Sjtim committed:
Split gradeDefinitionMutation into two seperat ones. One to activate / Deactivate and the other one to change Values.

Sjtim committed:
Added GradeView and PointsEdit components

Sjtim committed:
Db Changelog. Added Autograder Schema

Sjtim committed:
Added GradeDefinitionInput Field + less

Sjtim committed:
add GradeView + less

Sjtim committed:
add PointsEdit + less

Sjtim committed:
Modified Assignmentpage to fit Scoreboard. Added possibility to switch on/off Scoreboard

Sjtim committed:
Modified EditEvaluationPage to fit GradeDefinition Component.

Sjtim committed:
Added 3 Tasks for Evaluation

Sjtim committed:
Frontend Fixed:
- Grade-update on View
- Enhanced input Field for Float values
- adjusted layout for GradeDefinitionInput

Backend-Fixed
Race-Condition and Errors fixed.
Removed GradeCalculation in Autograding process. Requires a review where to exactly start it.
Grade Calculation currently only starts if a teacher adjusts just one point manually.

Overall Code Cleanup

Sjtim committed:
minor changes

Sjtim committed:
Bugs:
- fixed that a grade was not calculated after an evaluation.
- fixed race condition for gradecalculation.

Sjtim committed:
Bugs:
- GradeCalculation : missed an exit strategy if no gradecalculation is wanted (e. g. no autograder activated)

ts-linting done

Sjtim committed:
-performed gradle spotlessApply

Sjtim committed:
-Tasks removed

erikhofer commented:
Todo before review can start:
- [ ] Merge master and resolve conflicts
- [ ] Make sure that all checks pass
- [ ] Remove all commented out code

Sjtim committed:
Database validation per generateChangelog.sh bashscript

Sjtim committed:
- Changed variables names to make their purpose more clear.
- Added non-null assertion operators where it was required (linting)

Sjtim committed:
Merge remote-tracking branch 'origin/master' into autograder

# Conflicts:
#	client/package-lock.json
#	client/src/components/EvaluationResult.tsx
#	client/src/graphql/queries/evaluation.graphql
#	client/src/pages/assignment/AssignmentPage.tsx
#	client/src/pages/evaluation/EvaluationOverviewPage.tsx
#	src/main/kotlin/org/codefreak/codefreak/graphql/api/EvaluationApi.kt
#	src/main/kotlin/org/codefreak/codefreak/service/evaluation/EvaluationQueue.kt
#	src/main/kotlin/org/codefreak/codefreak/service/evaluation/EvaluationService.kt
#	src/main/resources/db/changelog-master.yaml
#	src/main/resources/init/tasks/codefreak.yml

Sjtim committed:
Merge branch 'master' into autograder

fwoelk commented on client/tsconfig.json:
```diff
@@ -18,8 +18,8 @@
     "isolatedModules": true,
     "noEmit": true,
     "jsx": "react-jsx",
-    "noUnusedLocals": true,
```
unused parameters and arguments make code a lot harder to understand. please undo this change

Sjtim committed:
- make the evaluationStatus available in their EvaluationResult children.

Sjtim committed:
- view added for the Grade

Sjtim committed:
- integrated PointsEdit view and a callback function to update the grade outside of this scope

Sjtim committed:
- update broke parser and formatter. Removed both and added an inputMode to numeric only

Sjtim committed:
- update broke parser and formatter. Removed both and added an inputMode to numeric only
- couple of possible null / undefined objects were caught with non-null assertion operator

Sjtim committed:
- couple of fixes required due to latest versions.
- numericSorter added but does not work properly?

Sjtim committed:
- fixed an issue which prevents a grade calculation
- removed unused code

Sjtim committed:
- removed unused code and an unreachable case.

Sjtim committed:
- removed unused code

Sjtim committed:
- restored

Sjtim committed:
- removed debounce in onChange event

Sjtim committed:
- unused variable removed and rearranged code

Sjtim committed:
- removed the unused PointsOfEvaluationStepDto which effected PointOfEvaluationStep updates and made the UI go bonkers on every value change.

Sjtim committed:
- missing import.

Sjtim committed:
- gradle spotlessApply
- npm run fix

Sjtim committed:
- fixed an architecture Violation.

Sjtim commented:
There is still an issue with sorting the scoreboard by grades. Im a bit lost with a proper solution which works. The Columns are generated by the amount of tasks and can contain nothing in case there is no evaluation / still in progress, an icon which says no grade is calculated due to an error or an actual grade.
Any ideas? See ../components/autograder/ScoreboardTable.tsx

Sjtim committed:
Merge branch 'master' into autograder

Sjtim committed:
- spotlessApply

Sjtim committed:
- fixed Sorter for Grade Columns.

Sjtim committed:
- Only Calc Points if the Grader is actually active if recalculation is required.

Sjtim committed:
- fixed an issue which prevented a grade calculation with the recent Changes on the Comment EvaluationStep. Since the Comments cant be removed in an evaluation the Autograder accepts one step to be not finished and has no EvaluationStepResult present.

Sjtim committed:
- SpotlessApply
- run fix

HenningCash commented on src/main/kotlin/org/codefreak/codefreak/service/evaluation/PointsOfEvaluationStepService.kt:
```diff
@@ -0,0 +1,237 @@
+   */
+  fun delete(poe: PointsOfEvaluationStep): Boolean {
+    return if (poeRepository.findById(poe.id).isPresent) {
+      poeRepository.delete(poe)
```
You can call this directly without checking for existence.

HenningCash commented on src/main/kotlin/org/codefreak/codefreak/service/evaluation/PointsOfEvaluationStepService.kt:
```diff
@@ -0,0 +1,237 @@
+      if (esService.updateResultFromPointsOfEvaluationStep(poe)) {
+        poe.calcCheck = true
+      } else {
+        LOG.error("Something went wrong in changeEvaluationStepStatus")
```
This won't be helpful when debugging

HenningCash commented on src/main/resources/evaluation/animals.json:
```diff
@@ -0,0 +1,226 @@
+[
```
As stated above. You could use a simple file where each line is a name

HenningCash commented on src/main/resources/init/tasks/codefreak.yml:
```diff
@@ -0,0 +1,8 @@
+title: "Sample Assignment"
```
Please remove `/init`

HenningCash reviewed (changes requested):
Some additional things:
Please review your code for unused parts and temporary changes you made that are not needed for autograding. I have not checked this for functional correctness (aka "behaves correctly"), yet.

**Frontend**:
* Please avoid `any` where possible
* Exessive `refetch` passing to child components can be prevented by using the `refetchQueries` option of mutations

**Backend**:
* Make sure every query/mutation is at least secured by `@Secured(Authority.ROLE_STUDENT)`
* Mutations need some form of authorization check if the user is allowed to edit this particular entity. This might also be true for some queries.
* Reduce logs to `debug` as much as possible. Mostly only unexpected errors are worth `info`

Sjtim committed:
- Removed EditNickname.less

Sjtim committed:
- fixed a case in which an evaluationStatus can be undefined.
- init grade = null
- retyped the status to evaluationStatus despite being from the EvaluationStepStatus Model. Otherwise its misleading.
- Added an interface to avoid fetchGrade from type any. Now it has its own interface.

Sjtim committed:
- removed Empty .less File.

Sjtim committed:
- removed init Task from Git-ignore. No longer supported.

Sjtim committed:
-removed init folder

Sjtim committed:
- shortened if-condition
- explanation why memorized Callback was used.

Sjtim committed:
- shortened if-condition

Sjtim committed:
- changed return type to null

Sjtim committed:
- shortened alias in a variable.
- shortened if-condition

erikhofer commented on src/main/kotlin/org/codefreak/codefreak/service/evaluation/GradeService.kt:
```diff
@@ -0,0 +1,189 @@
+  /**
+   * function to start a GradeCalculation
+   */
+  fun startGradeCalculation(eval: Evaluation): Grade? {
```
what does start mean? is it not finished within this function?

erikhofer commented on src/main/kotlin/org/codefreak/codefreak/service/evaluation/GradeService.kt:
```diff
@@ -0,0 +1,189 @@
+  /**
+   * Returns the best answer if one exists.
+   */
+  fun getBestGradeOfAnswer(answer: UUID): Grade? {
```
rename param to `answerId`

erikhofer commented on src/main/kotlin/org/codefreak/codefreak/service/evaluation/PointsOfEvaluationStepService.kt:
```diff
@@ -0,0 +1,237 @@
+@Service
+class PointsOfEvaluationStepService : BaseService() {
+
+  /**
+   * Logging
+   */
```
remove

erikhofer commented on src/main/kotlin/org/codefreak/codefreak/service/evaluation/PointsOfEvaluationStepService.kt:
```diff
@@ -0,0 +1,237 @@
+  @Autowired
+  private lateinit var gradeService: GradeService
+
+  /**
+   * Save Call of PointsOfEvaluationStep by EvaluationStepId
+   */
```
??

erikhofer commented on src/main/kotlin/org/codefreak/codefreak/service/evaluation/PointsOfEvaluationStepService.kt:
```diff
@@ -0,0 +1,237 @@
+    }
+  }
+
+  /**
+   * Get a save entity.
+   */
+  fun getEvaluationStepId(id: UUID): PointsOfEvaluationStep? {
```
remove comment. did you mean "by" id?

erikhofer commented on src/main/kotlin/org/codefreak/codefreak/service/evaluation/PointsOfEvaluationStepService.kt:
```diff
@@ -0,0 +1,237 @@
+    }
+  }
+
+  /**
+   * If Result was a success
+   */
+  fun onSuccess(es: EvaluationStep) {
```
remove comment. naming is a bit ambiguous. is this really public?

erikhofer commented on src/main/kotlin/org/codefreak/codefreak/service/evaluation/PointsOfEvaluationStepService.kt:
```diff
@@ -0,0 +1,237 @@
+   * starts calculation dependend of its associated EvaluationStepResult
+   */
+  fun calculate(es: EvaluationStep) {
+    LOG.info("calculate es: ${es.id}")
```
remove/debug

erikhofer commented on src/main/kotlin/org/codefreak/codefreak/service/evaluation/PointsOfEvaluationStepService.kt:
```diff
@@ -0,0 +1,237 @@
+    val gradeDefinition = es.gradeDefinition!!
+
+    val failedFeedbacks = feedbackRepository.findByEvaluationStepAndStatusNot(es, Feedback.Status.SUCCESS)
+    LOG.info("We gathered ${failedFeedbacks.size} Failed Feedbacks")
```
remove/debug

erikhofer commented on src/main/kotlin/org/codefreak/codefreak/service/evaluation/PointsOfEvaluationStepService.kt:
```diff
@@ -0,0 +1,237 @@
+
+    val mistakePoints = collectMistakes(failedFeedbacks, gradeDefinition)
+
+    LOG.info("we collected $mistakePoints mistakepoints")
```
remove/debug

erikhofer commented on src/main/kotlin/org/codefreak/codefreak/service/evaluation/PointsOfEvaluationStepService.kt:
```diff
@@ -0,0 +1,237 @@
+    }
+    poe.calcCheck = true
+    // print poe
+    LOG.info("just calculated: ${poe.reachedPoints} points! Did ${poe.mistakePoints} mistakes. CalcCheck is set to ${poe.calcCheck} ")
```
remove/debug

erikhofer commented on src/main/kotlin/org/codefreak/codefreak/service/evaluation/PointsOfEvaluationStepService.kt:
```diff
@@ -0,0 +1,237 @@
+    poe.calcCheck = true
+    // print poe
+    LOG.info("just calculated: ${poe.reachedPoints} points! Did ${poe.mistakePoints} mistakes. CalcCheck is set to ${poe.calcCheck} ")
+    // Save poe
```
remove

erikhofer commented on src/main/kotlin/org/codefreak/codefreak/service/evaluation/PointsOfEvaluationStepService.kt:
```diff
@@ -0,0 +1,237 @@
+      poe.reachedPoints = gradeDefinition.maxPoints - poe.mistakePoints
+    }
+    poe.calcCheck = true
+    // print poe
```
remove

erikhofer commented on src/main/kotlin/org/codefreak/codefreak/service/evaluation/PointsOfEvaluationStepService.kt:
```diff
@@ -0,0 +1,237 @@
+    poeRepository.save(poe)
+  }
+
+  /**
+   * If Result Errored
+   */
+  fun onErrored(es: EvaluationStep) {
+    // Nothings happen. Teacher has to take a look.
+  }
```
okay... is this needed?

erikhofer commented on src/main/kotlin/org/codefreak/codefreak/service/evaluation/PointsOfEvaluationStepService.kt:
```diff
@@ -0,0 +1,237 @@
+    }
+  }
+
+  /**
+   * CRUD-Functions
+   */
```
remove

erikhofer commented on src/main/kotlin/org/codefreak/codefreak/service/evaluation/PointsOfEvaluationStepService.kt:
```diff
@@ -0,0 +1,237 @@
+   * CRUD-Functions
+   */
+  fun findById(id: UUID): PointsOfEvaluationStep {
+    return poeRepository.findById(id).get()
```
EntityNotFoundException

erikhofer commented on src/main/kotlin/org/codefreak/codefreak/service/evaluation/PointsOfEvaluationStepService.kt:
```diff
@@ -0,0 +1,237 @@
+    }
+  }
+
+  /**
+   * Deletes a PointsOfEvaluationStep
+   */
```
remove

erikhofer reviewed (changes requested):
Nice work! I think the points of @HenningCash are valid. I've marked additional places where the code could use some finishing touches.

I don't fully understand how the autograding mechanism works yet, I think a few explaining comments would be helpful.

Also I think we should unify the naming of this feature. My suggestion would be to drop the term "autograder" and use "autograding" everywhere because this is a feature, not something with agency. What do you think @HenningCash ?

erikhofer commented:
Also there are no unit tests 🤔

Sjtim committed:
- Precious Application restored. :)

Sjtim committed:
- removed outcommented code

Sjtim committed:
- changed from json to simple text format

Sjtim committed:
- combined null and undefined check

Sjtim committed:
- removed outcommented code

Sjtim commented on client/src/components/autograder/GradeDefinitionInputField.tsx:
```diff
@@ -0,0 +1,194 @@
+    })
+  }, [gradeDefinition])
+
+  const memoizedCallback = useCallback(() => {
+    if (globalField !== null) {
+      if (input !== undefined)
+        debounce(updateMutation({ variables: { input } }).then, 500)
+    }
+  }, [globalField, updateMutation, input])
+
+  useEffect(() => {
+    memoizedCallback()
+  }, [gradeDefinition, memoizedCallback])
```
Whenever the Gradedefinition gets updated the useEffect Hook gets called and to prevent an infinite updateloop due to the gradeDefinition being in the dependency (tslint forced me to do this) i found a solution with memorizing the callback.

Sjtim commented on client/src/components/autograder/PointsEdit.tsx:
```diff
@@ -0,0 +1,157 @@
+
+const PointsEdit: React.FC<{
+  evaluationStepId: string
+  fetchGrade: any
```
evaluationStepId is used right below and fetchGrade as updatefunction.

Sjtim commented on client/src/hooks/useGetGrade.ts:
```diff
@@ -0,0 +1,39 @@
+    memorizeCallback()
+  }, [evaluationStepStatus, memorizeCallback])
+
+  function fetchGrade() {
```
Yes, this function gets passed down to the PointsEdit.tsx to update the Grade. This is needed because the teacher can modify points manually.

Sjtim commented on src/main/kotlin/org/codefreak/codefreak/CodeFreakApplication.kt:
```diff
@@ -16,16 +12,16 @@ import org.springframework.scheduling.annotation.EnableScheduling
-  override fun run(vararg args: String?) {
-    log.info("Code FREAK instance id: ${config.instanceId}")
-  }
+//  private val log = LoggerFactory.getLogger(this::class.java)
+//
+//  @Autowired
+//  private lateinit var config: AppConfiguration
+//
+//  override fun run(vararg args: String?) {
+//    log.info("Code FREAK instance id: ${config.instanceId}")
+//  }
```
Restored. :p

Sjtim commented on src/main/kotlin/org/codefreak/codefreak/entity/Assignment.kt:
```diff
@@ -48,6 +48,11 @@ class Assignment(
+  /**
+   * Let the teacher enable/disable the scoreboard
+   */
+  var scoreboard: Boolean = false
```
picked enableScoreboard. Henning was first. :D

Sjtim commented on src/main/kotlin/org/codefreak/codefreak/entity/GradeDefinition.kt:
```diff
@@ -0,0 +1,51 @@
+  /**
+   * Maximum of points reachable by its Grade definition
+   */
+  var maxPoints = 10f
```
maxPoints are anything but zero so i thought 10 is a good value?

Sjtim commented on src/main/kotlin/org/codefreak/codefreak/graphql/api/AssignmentApi.kt:
```diff
@@ -169,6 +174,18 @@ class AssignmentMutation : BaseResolver(), Mutation {
     true
   }
+
+  fun assignmentScoreboard(input: AssignmentScoreboardInputDto): Boolean = context {
```
May i ask why? Why do we want to update the whole assignment if just one value has changed?

erikhofer replied:
Because we don't want to clutter the API with mutations on a very granular level. All other entity update functions work this way so far.

Sjtim commented on src/main/kotlin/org/codefreak/codefreak/graphql/api/AutograderApi.kt:
```diff
@@ -0,0 +1,386 @@
+    val gradeDefinitionService = serviceAccess.getService(GradeDefinitionService::class)
+    val gradeDefinition = gradeDefinitionService.findGradeDefinition(input.id)
+
+    if (input.id == gradeDefinition.id) {
```
It actually cant. I picked it up somewhere. :D

Sjtim commented on client/src/components/autograder/GradeDefinitionInputField.tsx:
```diff
@@ -0,0 +1,194 @@
+  }
+
+  return (
+    <div className="grade-definition-input">
+      <div className="maxPointsInput">
```
Changed everything to the dashed conversion

Sjtim commented on client/src/components/autograder/ScoreboardTable.tsx:
```diff
@@ -0,0 +1,155 @@
+    // There should always be a grade defined
+    if (getAnswerFromSubmission(submission, task) !== undefined) {
+      const answer = getAnswerFromSubmission(submission, task)
+      // answer.
```
The comments? There is not always a grade actually. Removed. :p

Sjtim commented on client/src/components/autograder/ScoreboardTable.tsx:
```diff
@@ -0,0 +1,155 @@
+        key={`task-${task.id}`}
+        title={task.title}
+        align="center"
+        render={renderAnswer.bind((submission: any) => submission, task)}
```
Ive recycled this renderfunction from the SubmissionsTable. Put it back to undefined.

Sjtim commented on src/main/kotlin/org/codefreak/codefreak/entity/EvaluationStep.kt:
```diff
@@ -32,6 +33,12 @@ class EvaluationStep(
   var feedback = mutableSetOf<Feedback>()
 
+  @OneToOne(mappedBy = "evaluationStep", cascade = [CascadeType.ALL])
+  var pointsOfEvaluationStep: PointsOfEvaluationStep? = null
```
You want it to be named Points instead of PointsOfEvaluationStep? I could refactor it but Points could be anything in my opinion.

erikhofer replied:
Not the class but the field

Sjtim commented on src/main/kotlin/org/codefreak/codefreak/entity/EvaluationStep.kt:
```diff
@@ -32,6 +33,12 @@ class EvaluationStep(
+  var pointsOfEvaluationStep: PointsOfEvaluationStep? = null
+
+  @ManyToOne
+  var gradeDefinition: GradeDefinition? = null
```
It belongs to so many. :)
Is it an general design issue to keep the GradeDefinition foreign key in multiply Entities? I see why but i find long cascading harder to read. Thats why i keep the GradeDefinition in this case close.

erikhofer replied:
I mean "belong" in the sense of ownership. The relations of the ER model have semantics like "has a", "consists of" (aggregation, composition). One needs to have these relations in mind in order to make sense of the model. If we add addtional links on a technical level that don't have the same semantics, it is very confusing. We need to think about ownership and encapsulation. For you, it may be simpler to read if it is more consise but for everyone else it is harder to comprehend.

Sjtim commented on src/main/kotlin/org/codefreak/codefreak/entity/Grade.kt:
```diff
@@ -0,0 +1,45 @@
+   * A Grade is always part of a set in answer entity.
+   */
+  @ManyToOne
+  var answer: Answer
```
Every answer has a Grade. Otherwise every Evaluation had multiply grades and answers but which grade belongs to what answer?

erikhofer replied:
Every evalution belongs to a single answer && every grade belongs to a single evaluation => every grade belongs to a single answer. I don't see why we need an additional relation here 🤔

https://github.com/codefreak/codefreak/blob/f411cc31eecbb789ab071757c0c2135ed903ae8e/src/main/kotlin/org/codefreak/codefreak/entity/Evaluation.kt#L19

Sjtim replied:
Oh sorry, i mixed something up. Every Evaluation which has an Autograder turned on receives a grade. So if a student runs several evaluations on an answer there will be multiply evaluations and grades. We want to pick the best grade so i made a relation. Yes, i can access the grades through their evaluations. I thought it just would be simpler to access like i did with the GradeDefinition.
I will check the ER Model again and encapsulate everything properly. Didnt think it would be an issue.

In general i encountered a SQL-Deadlock in a previous project due to too much data and cascading. Kept several relations on fetchtype.LAZY and reduced the nessecity of especially long cascades i had to load everything manually, even unwanted entities, afterwards. Thats why i developed this foreign key behaviour. :p

erikhofer replied:
Ah, I see. Finding the best grade can be done on the database as we do with the _latest evaluation_
https://github.com/codefreak/codefreak/blob/7bd41e09b2eb005180f4e9a4f6fc4f43ce77553e/src/main/kotlin/org/codefreak/codefreak/repository/EvaluationRepository.kt#L11

In general I think this is primarily an issue of _premature optimization_. If there are indeed performance issues, I think they should be solved in another way. The same is true for lazy fetching of JPA entities. This is not always a good behavior and to turn it of we have to use custom JPQL queries.

I think you can leave it as is for now. @HenningCash and I will have a closer look on the ER model in general.

Sjtim replied:
Thats what i actually did in the GradeRepository.
` fun findFirstByAnswerIdOrderByGradePercentageDesc(id: UUID): Optional<Grade>`

Yes, i read the article you've posted. Insightful. :)
I'll keep the ER-Model as it is for now.

Sjtim commented on src/main/kotlin/org/codefreak/codefreak/entity/Grade.kt:
```diff
@@ -0,0 +1,45 @@
+   * Checks if a Grade is actually calculated. Otherwise a formatted zero might be shown in frontend due to a possible
+   * grade applied to their respective evaluation but no grade was calculated.
+   */
+  var calculated: Boolean = false
```
removed the variable calculated. Frameworks appear to be smarter than i thought. :D

Sjtim commented on src/main/kotlin/org/codefreak/codefreak/entity/Grade.kt:
```diff
@@ -0,0 +1,45 @@
+   * A Grade is calculated of multiply PointsOfEvaluationSteps.
+   */
+  @OneToMany(mappedBy = "grade", cascade = [CascadeType.ALL], orphanRemoval = true, fetch = FetchType.LAZY)
+  var pointsOfEvaluationStep = mutableSetOf<PointsOfEvaluationStep>()
```
Same as Answer to Grade. We would'nt be able to differ between what PointsOfEvaluationStep belongs to which Grade. This works for just one Evaluation but if a student sends in more it will become complicated. :p

Sjtim replied:
Mixed something up here. Yes they can get accessed through the Evaluation. Will Encapsulate.

Sjtim commented on src/main/kotlin/org/codefreak/codefreak/entity/GradeDefinition.kt:
```diff
@@ -0,0 +1,51 @@
+  /**
+   * Point reduction on minor errors
+   */
+  var minorError = 0f
+
+  /**
+   * Point reduction on major errors
+   */
+  var majorError = 0f
+
+  /**
+   * Point reduction on Critical Errors.
+   */
+  var criticalError = 0f
```
I actually changed them to minorSeverity, majorSeverity, criticalSeverity to make it clear to which fields they refer to.
Approved?

erikhofer replied:
No, that's still a non-descriptive name. The number is not "a severity". I think the comments are describing it correctly, so maybe `pointReductionForMinorMistakes` etc. Something like `minorMistakePenalty` would be more consise but lacks the unit of measurement. Still fine though i think.

Sjtim commented on src/main/kotlin/org/codefreak/codefreak/entity/PointsOfEvaluationStep.kt:
```diff
@@ -0,0 +1,70 @@
+  /**
+   * sum of all mistakes refered to all flawed feedbacks
+   */
+  var mistakePoints: Float = 0f
```
All Feedback which status is FAILED and severity is not INFO to be precise.

erikhofer replied:
so something like `totalPointReduction` ?

Sjtim commented on src/main/kotlin/org/codefreak/codefreak/graphql/api/AutograderApi.kt:
```diff
@@ -0,0 +1,386 @@
+ * GradeDefinitionMaxDto for query. Main purpose
+ */
+@GraphQLName("GradeDefinitionMax")
+class GradeDefinitionMaxDto(@GraphQLIgnore val entity: GradeDefinition, ctx: ResolverContext) : BaseDto(ctx) {
```
Its a short Version of a GradeDefinition which just provides the maxPoints to the frontend.
It is used in PointsEdit.tsx to display the relation between reached and max points.

erikhofer replied:
That's not needed. We use GraphQL, the frontend can query for a single field itself :)

Sjtim commented on src/main/kotlin/org/codefreak/codefreak/graphql/api/AutograderApi.kt:
```diff
@@ -0,0 +1,386 @@
+ * PointsOfEvaluation Dto for query
+ */
+@GraphQLName("PointsOfEvaluationStep")
+class PointsOfEvaluationStepDto(@GraphQLIgnore val entity: PointsOfEvaluationStep?, ctx: ResolverContext) : BaseDto(ctx) {
```
If there is no Autograding there wont be a PointsOfEvaluationStep entity.

erikhofer replied:
yeah but  then, why create a DTO at all? in that case the DTO may as well be null

Sjtim replied:
The DTO is null if there is no PointsOfEvaluationStep present, yes. This is the case if the Evaluation is running or no autograding is active. The EvaluationStepPanel.tsx contains a PointsEdit.tsx Field. If the DTO returns null, this field wont be shown. Otherwise Grading is active and it will be shown as simple view for students or edit field for teachers.
Same approach was done on the GradeDto.

Sjtim commented on src/main/kotlin/org/codefreak/codefreak/graphql/api/AutograderApi.kt:
```diff
@@ -0,0 +1,386 @@
+  constructor() : this(UUID.randomUUID(), 0f, 0f, 0f, 0f)
+}
+
+@GraphQLName("GradeDefinitionActiveInput")
+class GradeDefinitionActiveInputDto(var id: UUID, var active: Boolean) {
+  constructor() : this(UUID.randomUUID(), false)
+}
```
As you've mentioned this should be done via updateAssignment but why? Arent specialized mutations a good approach? Only pass the data we want to update?

erikhofer replied:
The idea of a GraphQL API is that it is generic. Clients need to be able to do what they want without having to chnage the API for each use case. Another solution would be a patch mutation that has all fields nullable and only values that are present are updated. But we don't have them yet. We may get to them at some point https://stackify.com/premature-optimization-evil/

Sjtim committed:
-refactored names

Sjtim committed:
-renamed scoreboard -> enableScoreboard to be consistent with backend

Sjtim committed:
- renamed gradeDefinition -> GradingDefinition and used as fragment

Sjtim committed:
- moved fragments and merged mutations / queries.

Sjtim committed:
In EvaluationHistory
- Ask for Authority in this scope. Any deeper implementation works only for latest and not previous Evaluations

In EvaluationResult
- passed authorization

In EvaluationStepPanel
- refactored code and passed authorization

In LatestEvaluation
- Ask for Authorization from Teacher and Pass it to the results.

Sjtim committed:
- changed MutableList to List

Sjtim committed:
- removed unused function

Sjtim committed:
- renamed GradeDefinition to GradingDefinition

Sjtim committed:
- renamed variable of PointsOfEvaluationStep and Refactored GradeDefinition to GradingDefinition

Sjtim committed:
- changed ER-Model and replaced the UserAlias Entity with a String.

Sjtim committed:
- removed own AssignmentScoreboard Query/Mutation and merged it into the existing updateAssignment.

Sjtim committed:
- removed GradingDefinitionActiveInput Query/Mutation and used the existing GradingDefinition instance to update the active attribute.

Sjtim committed:
- put Query into React Component and removed fetchScoreboard function which will now be handled on its child

Sjtim committed:
- renamed useGetGrade to useGrade and refactored the Hook.

Sjtim committed:
- refetches its own value after a mutation than the whole Query which was done previously.

Sjtim committed:
- fixed an issue after Query adjustments.

Sjtim committed:
- shortened the sorter for the task columns and refetches its own data instead of the whole query.

Sjtim committed:
- refactored and make more use of Hooks.
- added proper debounce implementation for user input
- In total runs much more smoothly

Sjtim committed:
- renamed to GradingDefinitionInputField
- refactored to make use of Hooks.
- added proper debounce function
- Followed recommendation and renamed attributs.

Sjtim committed:
- set Boolean required.

Sjtim committed:
- removed unused mutations

Sjtim committed:
- renamed GradeDefinition to GradingDefinition
- merged Queries / Mutations

Sjtim committed:
- renamed scoreboard to enableScoreboard like recommended

Sjtim committed:
- Added missing datatypes and removed relation to PointsOfEvaluationStep

Sjtim committed:
- removed unnessecary relation and as recommended removed the calculated variable and made the gradePercentage nullable

Sjtim committed:
- renamed variables and made comments more clear.
- removed relation to GradingDefinition. Reachable through EvaluationStep.

Sjtim committed:
- renamed GradeDefinition to GradingDefinition
- Useralias Entity removed

Sjtim committed:
- refactored conditions for autograding. See comments

Sjtim committed:
- removed unused functions and added EntityNotFoundExceptions

Sjtim committed:
- refactored the Points calculation process and added comments for understanding

Sjtim committed:
- refactored the calculation process and added comments. Also removed unused code / bad comments / unused functions.

Sjtim committed:
- Changed the way an alias name is implemented. It is no longer an own Entity rather a String attribute in the User Entity.

Sjtim committed:
- spotlessApply

Sjtim committed:
- Adjusted Repository. Removed unused functions and added new ones. UserAliasRepository is deleted.

Sjtim committed:
- Database Schema adjustments

Sjtim committed:
Merge pull request #1 from codefreak/master

Updated

Sjtim committed:
Merge branch 'master' into autograder

# Conflicts:
#	src/main/kotlin/org/codefreak/codefreak/service/evaluation/EvaluationQueue.kt

Sjtim committed:
- showed a false label instead of nothing

Sjtim committed:
- fixed a status which was not considered and threw an error.

Sjtim committed:
- renamed auth to authority
- considered the EvaluationStatus as undefined.

Sjtim committed:
- refactored to solve an infinite loop bug.

Sjtim committed:
- enableScoreboard optional value again.
- added a mutation to update GradingDefinitionValues. Its the possibility in passing an InputDto instead of single values.

Sjtim committed:
- enableScoreboard optional with Boolean?

Sjtim committed:
- removed unused functions

Sjtim committed:
wording

Sjtim committed:
- removed hotfix

Sjtim committed:
- reworked and added comments

Sjtim committed:
- added missing grade calculation on GradingDefinition mutation.

Sjtim committed:
- refactored code and added / changed comments

Sjtim committed:
- enableScoreboard modified.

Sjtim committed:
- removed unused function
- Grade is optional and is null if not existing.

Sjtim committed:
- removed questionmark. Will be false by default.

Sjtim committed:
- spotlessApply / npm run fix

Sjtim committed:
- Float4 -> Float
